akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -746,10 +746,50 @@ private[akka] class ShardRegion(
 
   var coordinator: Option[ActorRef] = None
 
+  def reRegisterIfCoordinatorNotUp(): Unit =
+    if (coordinator.nonEmpty) {
+      val coordAddress = coordinator.get.path.address // safe: guarded by nonEmpty
+      val coordinatorStatus =
+        coordinator.flatMap { _ =>
+          membersByAge.find(_.address == coordAddress).map(_.status)
+        }
+
+      coordinatorStatus match {
+        case Some(MemberStatus.Up) => () // Do nothing
+        case Some(notUp) =>
+          if (log.isDebugEnabled)
+            log.debug(
+              "{}: Coordinator is on node with status [{}], proactively attempting to find next coordinator",
+              typeName,
+              notUp)
+
+          // For now, attempt to register with the oldest Up member we know about without forgetting
+          // about the current coordinator
+          // We only attempt one candidate so as to not flood with registration messages
+          // Important since this is level-triggered (any membership change where the coordinator is on a
+          // not-up node) while registration is otherwise edge-triggered
+          coordinatorSelection.headOption.foreach(sendRegistrationMessage)
+
+          // in case we're not getting any membership changes for a while...
+          if (!timers.isTimerActive(RegisterRetry)) {
```

Comment:
will this work as intended, since we don't set `coordinator = None` for this case?

```
case RegisterRetry =>
      if (coordinator.isEmpty) {
```

Reply:
we watch the current coordinator, and when it terminates we will `startRegistration()` so that is already covered

Reply:
`RegisterRetry` is also changed to `reRegisterIfCoordinatorNotUp`.

Yes, eventually the current coordinator will stop (worst case, thanks to the failure detector before removal gossip) and our watch will trigger full registration.  That can be several seconds in the future.

Reply:
Scheduling the retry also helps in the event that the next-youngest is also being stopped by (e.g.) kubernetes (consider a deployment where `maxSurge`/`maxUnavailable` allows multiple pods to be stopped per round and deletion cost is being set based on cluster age: then it's likely that the oldest `n` pods get stopped) but we haven't found out about it yet.

An alternative could be to compute the youngest `Up` on membership changes and go through this if that changes, but I think a fairly quick (default 250ms) retry is sufficient?

Reply:
I think we are good here. Should also be fine to configure a shorter retry-interval. Default is 2 seconds. The initial interval is derived from that, but at least 100 ms.